### PR TITLE
Fix/318/fix invalid semantic versions

### PR DESCRIPTION
**Pull request by IainSAP**
Fix for: https://github.com/SAP/open-ux-tools/issues/318

Adds proper version -> semantic version conversion so that we no longer write invalid versions to yamls and manifest.json. Doing so was causing failures to load UI5 libraries.

Also, adds localisation strings for `ui5-application-writer`.

Updates ui5 version -> manifest schema version from https://github.com/SAP/ui5-manifest/blob/master/mapping.json

Updates to properly handle snapshot versions when comparing versions and looking up manifest descriptor version.

### CODE REVIEW

**Review by zdravko-georgiev — DISMISSED** [review]
Thanks @IainSAP

File: packages/ui5-application-writer/src/data/validators.ts [hunk 1]
```diff
@@ -9,11 +11,27 @@ import { Ui5App } from '../types';
  */
 export function validateAppId(appId: string): boolean {
     if (!appId) {
-        throw new Error('The property: app.id must have a value'); // todo: add localised parameterised message
+        throw new Error(t('error.missingRequiredProperty', { propertyName: 'app.id' }));
```
Comment (IainSAP) [inline packages/ui5-application-writer/src/data/validators.ts:14]: Test needed to cover : https://sonarcloud.io/code?id=SAP_open-ux-tools&pullRequest=319&selected=SAP_open-ux-tools%3Apackages%2Fui5-application-writer%2Fsrc%2Fdata%2Fvalidators.ts&line=33

**Comment by IainSAP** [comment]
> @IainSAP changes look good. But can you add tests for UI5 snapshot versions. E.g. `snapshot-1.98` `semVer.coerce` might remove `snapshot-` in the minUI5Version incorrectly.

Yes it will remove that, currently adding tests.

File: packages/ui5-application-writer/test/data.test.ts [hunk 2]
```diff
@@ -113,6 +113,42 @@ describe('Setting defaults', () => {
...
+        {
+            input: {
+                version: 'snapshot-1.80',
+                frameworkUrl: 'https://ui5.unreleased.sap.com'
+            },
+            expected: {
+                framework: 'SAPUI5',
+                frameworkUrl: 'https://ui5.unreleased.sap.com',
+                version: 'snapshot-1.80',
+                localVersion: 'snapshot-1.80',
```
Comment (devinea) [inline packages/ui5-application-writer/test/data.test.ts:126]: Local version should be `1.80.0`

File: packages/ui5-application-writer/test/data.test.ts [hunk 3]
```diff
@@ -113,6 +113,42 @@ describe('Setting defaults', () => {
...
+                minUI5Version: '1.60.0',
```
Comment (devinea) [inline packages/ui5-application-writer/test/data.test.ts:127]: minUI5Version should be ‘snapshot-1.80’
@zdravko-georgiev correct me if i am wrong.

File: packages/ui5-application-writer/src/data/defaults.ts [hunk 4]
```diff
@@ -75,19 +80,41 @@ export function mergeUi5(ui5: Partial<UI5>): UI5 {
         framework,
         frameworkUrl: ui5.frameworkUrl ?? defaultFrameworkUrl
     };
-    const typesVersion = parseFloat(merged.localVersion) >= 1.76 ? merged.localVersion : '1.71.18';
+    // typesVersion must be a valid npm semantic version, we know they cannot be null as already validated
+    const localSemVer = semVer.valid((semVer.coerce(merged.localVersion)))!;
+    const typesVersion = semVer.gte(localSemVer, '1.76.0') ? localSemVer : '1.71.18';
```
Comment (ullasholla) [inline packages/ui5-application-writer/src/data/defaults.ts:85]: Use constants for `'1.76.0'` and `'1.71.18'` too? I know this PR does not add this code but touches it.

File: packages/ui5-application-writer/src/data/defaults.ts [hunk 5]
```diff
@@ -113,8 +140,13 @@ function getLocalVersion({
...
         }
-        if (parseFloat(version) > parseFloat(result)) {
-            result = version;
+
+        if (version !== UI5_DEFAULT.DEFAULT_UI5_VERSION) {
+
+            // Update to a valid coerced version string e.g. 1.80 -> 1.80.0. Cannot be null as previously validated.
+            if (semVer.gt(semVer.coerce(version)!, semVer.coerce(result)!)) {
+                result = version;
+            }
```
Comment (ullasholla) [inline packages/ui5-application-writer/src/data/defaults.ts:149]: Is the `else` before the addition (line 140) the same as the this condition: `if (version !== UI5_DEFAULT.DEFAULT_UI5_VERSION)`? If so, there seems to be a chance an open UI5 version could get overwritten by the addition.

File: packages/ui5-application-writer/src/i18n.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Helper function facading the call call to i18next.
+ *
+ * @param key i18n key
+ * @param options additional options
+ * @returns {string} localized string stored for the given key
+ */
+export function t(key: string, options?: TOptions): string {
+    return i18next.t(key, options);
+}
```
Comment (ullasholla) [inline packages/ui5-application-writer/src/i18n.ts:32]: Can we directly use the `i18next.t()` here instead of wrapping it?

**Review by ullasholla — CHANGES_REQUESTED** [review]
Thanks for the fix. Code looks good. Test coverage is good. Using `semver` is a good idea.

Have a couple of questions? Could you take a look?

File: packages/ui5-application-writer/test/data.test.ts (hunk 3 above)
Comment (IainSAP) [reply packages/ui5-application-writer/test/data.test.ts:127]: @devinea This is how it was implemented...if `minUI5Version` is not provided it defaults to `1.60` https://github.com/SAP/open-ux-tools/pull/319/files#diff-0f207248944a1732a5b7343b9d1bb29e22bbae5fa2acce7a573a344c04c856d2L72

But I think you are correct, I will update to use the current version provided if not present.

File: packages/ui5-application-writer/src/data/defaults.ts (hunk 5 above)
Comment (IainSAP) [reply packages/ui5-application-writer/src/data/defaults.ts:149]: I believe `result` would be overwritten by the addition, at least that's how it was previously implemented. But I see we can refactor here to make the intention clearer.

File: packages/ui5-application-writer/test/data.test.ts (hunk 2 above)
Comment (zdravko-georgiev) [reply packages/ui5-application-writer/test/data.test.ts:126]: @devinea @IainSAP What was the `localVersion` again? The version for the `ui5-local.yaml`? If yes, then `localVersion` can't be a snapshot. Only released UI5 versions 1.76.0 and above.

File: packages/ui5-application-writer/test/data.test.ts (hunk 3 above)
Comment (zdravko-georgiev) [reply packages/ui5-application-writer/test/data.test.ts:127]: @devinea @IainSAP We need to be careful here, If I am not mistaken snapshot versions are not valid for the `manifest.json`.

File: packages/ui5-application-writer/src/i18n.ts (hunk 6 above)
Comment (IainSAP) [reply packages/ui5-application-writer/src/i18n.ts:32]: I believe it's done like this due ordering of `i18next.init` on load of the file and we don't want consumers to have to manage that process. Also this is replicated code across many modules and should be consolidated to a single i18n utility in another TBI and PR so I will leave it as is for now.

File: packages/ui5-application-writer/test/data.test.ts (hunk 2 above)
Comment (devinea) [reply packages/ui5-application-writer/test/data.test.ts]: @zdravko-georgiev  Yeah `localVersion` is the one for `ui5-local.yaml`
@IainSAP  - Okay yeah I guess if `localVersion` is passed in as `snapshot-1.80` it should be output the same. Even if it is not really valid for the ui5-local.yaml.

File: packages/ui5-application-writer/src/data/version-to-descriptor-mapping.json [hunk 7]
```diff
@@ -1,5 +1,10 @@
 {
-    "latest": "1.33.0",
+    "latest": "1.38.0",
+    "1.98": "1.38.0",
+    "1.96": "1.37.0",
+    "1.95": "1.36.0",
+    "1.93": "1.35.0",
+    "1.92": "1.34.0",
```
Comment (devinea) [inline packages/ui5-application-writer/src/data/version-to-descriptor-mapping.json:7]: Thanks

**Review by devinea — APPROVED** [review]
Thanks @IainSAP

**Comment by IainSAP** [comment]
Thx all

**Pull request merged**